Bugfix [Content sharing opt] FXIOS-11269 Postpone tab restoration when handling deeplinks

FilippoZazzeroni opened the pull request:
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-11269)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/24513)

## :bulb: Description
Remove tab restoration and startup flow complete filter when handling a deeplink and postpone tab restoration after deeplink as been handled.
The deeplink tab is appended to the restore tabs after restoration or restored is content if it was already present in the previous section, in order to maintain the previous behavior.
Add `deeplinkOptimizationRefactor` feature flag and related tests. 

## :pencil: Checklist
You have to check all boxes before merging
- [ ] Filled in the above information (tickets numbers and description of your work)
- [ ] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

FilippoZazzeroni commented:
Hii 🔆 i opened this as draft PR since the tab implementation is still WIP, but i'd like to gather some feedback from you mostly regarding tab work. I think this implementation doesn't generate any data race conditions since it doesn't put restoration in parallel with deeplinks. But i may forgetting something. Also @Cramsden for the tab now i just added a check if we have already a tab there that would be the deeplink tab, then continue the restoration process. I don't like this, what do you think if we refactor it or we treat the `DeeplinkTab` differently and instead of relying on count we make or a separate object or something that can identifies it. This would be also beneficial, if we want to move that tab around, since with this implementation the Deeplink tab goes first in the list, and i think it is ok as a behavior, but we might want or need to change it.

mattreaganmozilla commented on firefox-ios/Client/Application/SceneDelegate.swift:
```diff
@@ -234,11 +234,6 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
         logger.log("Scene coordinator will handle a route", level: .info, category: .coordinator)
         sessionManager.launchSessionProvider.openedFromExternalSource = true
 
-        AppEventQueue.wait(for: [.startupFlowComplete, .tabRestoration(sceneCoordinator.windowUUID)]) { [weak self] in
-            self?.logger.log("Start up flow and restoration done, will handle route",
-                             level: .info,
-                             category: .coordinator)
-            sceneCoordinator.findAndHandle(route: route)
-        }
```
Thanks @FilippoZazzeroni for investigating this! This change makes me a little nervous, but perhaps I'm just misunderstanding the big-picture changes. Why would we not want to wait for startup completion & tab restoration before attempting to handle the route? My recollection is that we can't safely open a new tab from a deeplink until both of those are completed. Is that not correct?

FilippoZazzeroni replied:
Hi @mattreaganmozilla the scope of this would be to optimize the deeplink load time, since right now is quite slow. 
Syncing up with @lmarceau told me that the filter was in place to avoid data race, since the deeplink could be loaded while the tab restoration process was running, leading to uncertain results. This changes doesn't try to parallelize both but to postpone restoration process after deeplink handling so to prioritize it. It should be safe, this removal, but i'm still investigating potential drawbacks. I may missing still something, let me know if you have any test i could run to test this more accurately. Thanks @mattreaganmozilla 😃

FilippoZazzeroni commented:
The PR is almost ready for review. I added a feature flag to revert the implementation. I still need to investigate a reliable way to cancel startup record for all the deeplink possibilities and update tests. 
@mattreaganmozilla , @Cramsden the tab manager now stores the deeplink tab, and updates its position or content during the restore process, let me know if this implementation works out. Thanks so much 😃 🌞

FilippoZazzeroni commented:
@lmarceau @Cramsden @mattreaganmozilla I marked the PR ready for review. I added tests to `TabManagerTests` for the restoreTab and restored the previous once.
I did refactor a bit the code to ensure the complete action is sent correctly and for all the possible deeplinks situations.

lmarceau commented on firefox-ios/nimbus-features/deeplinkOptimizationRefactorFeature.yaml:
```diff
@@ -0,0 +1,18 @@
+        description: >
+          Variable description
+        type: Boolean
+        default: false
+    defaults:
+      - channel: beta
+        value:
+          enabled: false
```
Do we want to do a rollout for this change? Maybe we should have it ON in beta too? Testing with Testflight is a necessary step for sure since deeplinks can behave differently with application that have been put in background for a while

lmarceau commented on firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:
```diff
@@ -2610,33 +2623,26 @@ class BrowserViewController: UIViewController,
         } else if let legacyUrlBar {
             urlBar(legacyUrlBar, didSubmitText: query)
         }
-        AppEventQueue.signal(event: .recordStartupTimeOpenURLComplete)
```
Same, we don't want this here anymore?

FilippoZazzeroni replied:
I moved this in the `SceneDelegate` so there i catch all the possible deeplinks cases.

FilippoZazzeroni commented on firefox-ios/Client/TabManagement/Tab.swift:
```diff
@@ -1267,6 +1267,7 @@ class TabWebView: WKWebView, MenuHelperWebViewInterface, ThemeApplicable {
 
     func addPullRefresh(onReload: @escaping () -> Void) {
         guard !scrollView.isZooming else { return }
+        guard scrollView.frame != .zero else { return }
```
That is need because when handling deeplinks the tab is added before the scrollView has a frame, so in case the frame is zero i don't create a pull request, that will be added after the page loads.

lmarceau replied:
And that change doesn't need to be protected by a feature flag I guess?

thatswinnie replied:
Let's also combine the two `guard` statements if possible.

lmarceau reviewed (approved):
I'm approving now since I am leaving for PTO but I think we should get another approval before moving forward with this. Since this is feature flag protected, I feel like it's ok to make this more risky change. We should ensure QA thoroughly test deeplinks in Beta (specifying long time backgrounding the application). Maybe we need a QA request for that (we can check with Andy!). Great work @FilippoZazzeroni !

FilippoZazzeroni commented:
Thanks @lmarceau i agree this should have at least another review, since is touching a core part of the app. 
cc @mattreaganmozilla, @OrlaM can you please have a look to this PR ?

mattreaganmozilla commented on firefox-ios/Client/Application/SceneDelegate.swift:
```diff
@@ -20,6 +22,9 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
     private let logger: Logger = DefaultLogger.shared
     private let tabErrorTelemetryHelper = TabErrorTelemetryHelper.shared
+    private var isDeeplinkOptimizationRefactorEnabled: Bool {
+        return featureFlags.isFeatureEnabled(.deeplinkOptimizationRefactor, checking: .buildOnly)
```
Is there any issue if this flag changes during the runtime/lifecycle of the app? (IOW should we check this once and cache the value for that app lifecycle, or hit `isFeatureEnabled` every time as it currently is?)

FilippoZazzeroni replied:
Theoretically this @mattreaganmozilla cannot change during the life time of the app, i guess. I don't know if the enrollment can take place during the app lifecycle, but i believe no. i'm going to check this further, thanks for bringing this up. 😃

OrlaM replied:
Flags don't currently change during the apps life, a restart is needed.

mattreaganmozilla commented on firefox-ios/Client/TabManagement/TabManagerImplementation.swift:
```diff
@@ -161,6 +167,10 @@ class TabManagerImplementation: NSObject, TabManager, FeatureFlaggable, TabEvent
     }
 
+    deinit {
+        notificationCenter.removeObserver(self)
```
I don't think this is necessary? Doc snippet:

> If your app targets iOS 9.0 and later or macOS 10.11 and later, and you used [addObserver(_:selector:name:object:)](https://developer.apple.com/documentation/foundation/notificationcenter/1415360-addobserver), you do not need to unregister the observer. 

Looks like the observation occurs in `Notfiable.setupNotifications` which appears to use that function but LMK if I'm missing something however.

mattreaganmozilla reviewed (commented):
> cc @mattreaganmozilla, @OrlaM can you please have a look to this PR ?

I took a quick look at the changes (nice work on this @FilippoZazzeroni). I haven't had time to do a deep dive of this PR though or test the branch yet. The change seems risky to me just because this area of the codebase is somewhat brittle, but if it's behind a feature flag then I'm fine with us moving forward. Maybe @OrlaM will have time to lend her 👀 to this.

FilippoZazzeroni commented:
@OrlaM , @mattreaganmozilla i moved the enqueue of the tab restoration to `viewDidLoad` and added an additional check to the signal in order to avoid calling restoreTabs more the once. @mattreaganmozilla if you think there is a better approach with `AppEventQueue` please let me know, i couldn't find a better way that the one i provided. Thanks for the reviews 😃 . Hopefully this PR can land in 137, since this issue is currently a top one reported by users.

thatswinnie commented on firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:
```diff
@@ -1000,11 +1005,34 @@ class BrowserViewController: UIViewController,
     }
 
+    private func enqueueTabRestoration() {
+        if isDeeplinkOptimizationRefactorEnabled {
```
Can we use `guard` here instead?

thatswinnie commented on firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:
```diff
@@ -1000,11 +1005,34 @@ class BrowserViewController: UIViewController,
+                    self?.tabManager.restoreTabs()
+                }
+            }
+            if !AppEventQueue.hasSignalled(.recordStartupTimeOpenDeeplinkComplete) {
```
Can we use `else if` here?

thatswinnie commented on firefox-ios/Client/TabManagement/TabManagerImplementation.swift:
```diff
@@ -696,7 +687,62 @@ class TabManagerImplementation: NSObject, TabManager, FeatureFlaggable, TabEvent
+        let newTab: Tab
+        if isDeeplinkOptimizationRefactorEnabled {
+            if let deeplinkTab, deeplinkTab.url?.absoluteString == tabData.siteUrl {
+                newTab = deeplinkTab
```
Could you please add a comment here?

FilippoZazzeroni replied:
yes sure i'm adding it
